### PR TITLE
APP-421 - Background activity rework

### PR DESCRIPTION
**Pull request by wizzard**
Implemented 2 polling interfaces:
* `curl_multi_wait()` for newer cURL library versions (added in cURL 7.28.0)
* `select()` as a fallback for older cURL library versions

### CODE REVIEW

File: src/background_activity.c [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+    curl_h *head;
+    curl_h *tail;
+};
+
+struct curl_h {
+    CURL *curl;
+    int id;
+    curl_h *prev;
+    curl_h *next;
+    cqueue_t *owner;
```
Comment (nitzanpx) [inline src/background_activity.c]: can you explain why we need to know about `cqueue_t` from the curl_h?

File: src/background_activity.c [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, thd_data->server, LOGGER_ERROR_HDR"Failed to call apr_pool_create()", conf->app_id);
+        return NULL;
+    }
+
+    // TODO: app_id can be updated via remote_config
+    app_id = apr_pstrdup(pool, conf->app_id);
+
+    rv = cqueue_create(&q_waiting, pool);
+    if (rv != APR_SUCCESS) {
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, thd_data->server, LOGGER_ERROR_HDR"Failed to call apr_pool_create()"LOGGER_FTR);
```
Comment (nitzanpx) [inline src/background_activity.c]: wrong error msg

File: src/background_activity.c [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        return NULL;
+    }
+    rv = cqueue_create(&q_busy, pool);
+    if (rv != APR_SUCCESS) {
+        ap_log_error(APLOG_MARK, APLOG_ERR, 0, thd_data->server, LOGGER_ERROR_HDR"Failed to call apr_pool_create()"LOGGER_FTR);
```
Comment (nitzanpx) [inline src/background_activity.c]: wrong error msg

File: src/background_activity.c [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+
+    // init curl_h array, put in the waiting queue
+    curl_h **cha = apr_pcalloc(pool, sizeof(curl_h *) * conf->background_activity_workers);
+    for (i = 0; i < conf->background_activity_workers; i++) {
+        curl_h *ch = apr_pcalloc(pool, sizeof(curl_h));
+        cha[i] = ch;
+        ch->id = i;
+        ch->curl = curl_easy_init();
+        if (!ch->curl) {
+            ap_log_error(APLOG_MARK, APLOG_ERR, 0, thd_data->server, LOGGER_ERROR_HDR"background_activity: failed to create cUCL handle"LOGGER_FTR);
```
Comment (nitzanpx) [inline src/background_activity.c]: `cURL` instead of `cUCL`

File: src/background_activity.c [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+                cqueue_push(q_busy, ch);
+
+                // assuming to be a zero terminated string
+                curl_easy_setopt(ch->curl, CURLOPT_COPYPOSTFIELDS, activity);
+
+                // conf->activities_api_url can be modified by remote_config thread
+                if (conf->remote_config_enabled) {
+                    apr_thread_rwlock_rdlock(conf->remote_config_lock);
+                    url = apr_pstrdup(pool, conf->activities_api_url);
+                    apr_thread_rwlock_rdlock(conf->remote_config_lock);
```
Comment (nitzanpx) [inline src/background_activity.c]: did you want to call unlcok?

File: src/background_activity.c [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+                        LOGGER_DEBUG_HDR"background_activity: failed to extract item from the waiting queue"LOGGER_FTR);
+                    free(activity);
+                    continue;
+                }
+                cqueue_push(q_busy, ch);
+
+                // assuming to be a zero terminated string
+                curl_easy_setopt(ch->curl, CURLOPT_COPYPOSTFIELDS, activity);
+
+                // conf->activities_api_url can be modified by remote_config thread
```
Comment (nitzanpx) [inline src/background_activity.c]: In general, I dont think we will let a user to change url apis

File: src/background_activity.c [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+#else
+        // wait until one of fd is active or timeout
+        res = curl_multi_wait(multi_curl, waitfd, 1, SELECT_TIMEOUT_MS, &numfds);
+        if (res != CURLM_OK) {
+            ap_log_error(APLOG_MARK, APLOG_ERR, 0, thd_data->server, LOGGER_ERROR_HDR"background_activity: curl_multi_wait() error: %d"LOGGER_FTR, res);
+        }
+
+        // pipe is readable
+        if (numfds && waitfd[0].revents & CURL_WAIT_POLLIN) {
+            if (wakeup_read(conf->background_activity_wakeup_fds) != APR_SUCCESS) {
```
Comment (nitzanpx) [inline src/background_activity.c]: please elaborate why we need to wait until wakeup again? we also have this code at the beginning?

File: src/background_activity.c [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+        // wait until one of fd is active or timeout
+        if (select(maxfd+1, &rfds, &wfds, &efds, &wait) < 0) {
+
+            // something bad
+            // TODO: handle select error
+            ap_log_error(APLOG_MARK, APLOG_ERR, 0, thd_data->server, LOGGER_ERROR_HDR"background_activity: select error: %s"LOGGER_FTR, strerror(errno));
+
+            // reset
+            FD_ZERO(&rfds);
+            continue;
```
Comment (nitzanpx) [inline src/background_activity.c]: shouldn't we have a continue in the ifndef else as well?

File: src/background_activity.c (hunk 1 above)
Comment (wizzard) [reply src/background_activity.c]: In our code, `curl_h` could belong to one of two `cqueue_t`: the queue of "waiting" handles and the queue of "busy" handles. In `cqueue_remove()` function have a check to make sure we remove a handle from the right queue, so we have to keep a pointer to the owner `cqueue_t` in `curl_h` structure.

File: src/background_activity.c (hunk 7 above)
Comment (wizzard) [reply src/background_activity.c]: We have 2 separate code paths here, separated by `#ifdef`:
* if a system has a "modern" cURL with `curl_multi_wait`: then we perform `wakeup_read()` at this point
* if cURL library does not have `curl_multi_wait`: then we perform `wakeup_read()` at the beginning of the loop

File: src/background_activity.c (hunk 1 above)
Comment (nitzanpx) [reply src/background_activity.c]: 👍

File: src/background_activity.c (hunk 7 above)
Comment (nitzanpx) [reply src/background_activity.c]: 👍

File: src/mod_perimeterx.c [hunk 9]
```diff
@@ -348,9 +350,18 @@ int px_handle_request(request_rec *r, px_config *conf) {
...
+
+    // thread-owned pool
+    rv = apr_pool_create(&pool, NULL);
+    if (rv != APR_SUCCESS) {
+        ap_log_error(APLOG_MARK, APLOG_CRIT, rv, hc->server, "error while trying to initialize apr_pool");
+        return NULL;
+    }
 
-    const char *health_check_url = apr_pstrcat(hc->server->process->pool, hc->config->base_url, HEALTH_CHECK_API, NULL);
+    const char *health_check_url = apr_pstrcat(pool, hc->conf->base_url, HEALTH_CHECK_API, NULL);
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: we can use here conf->base_url

File: src/mod_perimeterx.c [hunk 10]
```diff
@@ -348,9 +350,18 @@ int px_handle_request(request_rec *r, px_config *conf) {
...
     thread_data *hc = (thread_data*) data;
-    px_config *conf = hc->config;
+    px_config *conf = hc->conf;
+    apr_pool_t *pool;
+    apr_status_t rv;
+
+    // thread-owned pool
+    rv = apr_pool_create(&pool, NULL);
+    if (rv != APR_SUCCESS) {
+        ap_log_error(APLOG_MARK, APLOG_CRIT, rv, hc->server, "error while trying to initialize apr_pool");
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: can we change this to new debug format?

File: src/mod_perimeterx.c [hunk 11]
```diff
@@ -381,34 +393,65 @@ static void *APR_THREAD_FUNC health_check(apr_thread_t *thd, void *data) {
...
-    px_config *conf = th_data->config;
+    px_config *conf = th_data->conf;
     bool startup = true;
 
     apr_pool_t *pool;
     apr_status_t rv;
     rv = apr_pool_create(&pool, conf->pool);
     if (rv != APR_SUCCESS) {
-        ap_log_error(APLOG_MARK, APLOG_CRIT, rv, th_data->server, LOGGER_ERROR_FORMAT, conf->app_id, "remote configuration: error while trying to initialize apr_pool");
+        ap_log_error(APLOG_MARK, APLOG_CRIT, rv, th_data->server, "error while trying to initialize apr_pool");
```
Comment (nitzanpx) [inline src/mod_perimeterx.c]: replace ap_log_error to new debug format

File: src/mod_perimeterx.c (hunk 10 above)
Comment (wizzard) [reply src/mod_perimeterx.c]: we can't, as the new debug functions use `pool` variable to allocate temporary memory, which is not available in this line of code.

File: src/mod_perimeterx.c (hunk 11 above)
Comment (wizzard) [reply src/mod_perimeterx.c]: the same here: `pool` variable is not available, we can not use the new debug functions.

File: src/px_payload.c [hunk 12]
```diff
@@ -165,10 +167,11 @@ risk_payload *parse_risk_payload(const char *raw_payload, request_context *ctx)
     return rp;
 }
 
-static int digest_payload1(const risk_payload*payload, request_context *ctx, const char *payload_key, const char **signing_fields, char *buffer, int buffer_len) {
-
+static int digest_payload1(const risk_payload *payload, request_context *ctx, const char *payload_key, const char **signing_fields, char *buffer, int buffer_len) {
     unsigned char hash[32];
 
+    (void)ctx;
```
Comment (nitzanpx) [inline src/px_payload.c]: if we're not using this variable we can remove it

Comment (nitzanpx) [reply src/px_payload.c]: are we using it in `digest_payload3`

Comment (wizzard) [reply src/px_payload.c]: Let's leave `ctx` for now, as one of the use of `request_context *ctx` is to pass `conf` variable into functions, we use `conf` to call the new logging functions. I'd like to add additional logging to the cases where HMAC_* functions return failure.

**Pull request merged**